### PR TITLE
#WAF Adding WAF support to PowerShell

### PR DESCRIPTION
**Pull request by jobatzil**
## Description
Add support of newest WAF functionality to PowerShell.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ ] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/Compute/Commands.Compute.Test/Commands.Compute.Test.csproj [hunk 1]
```diff
@@ -67,8 +67,8 @@
       <HintPath>..\..\..\packages\Microsoft.Azure.Management.Compute.14.0.0-prerelease\lib\net45\Microsoft.Azure.Management.Compute.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="Microsoft.Azure.Management.Network, Version=9.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.9.1.0-preview\lib\net45\Microsoft.Azure.Management.Network.dll</HintPath>
+    <Reference Include="Microsoft.Azure.Management.Network, Version=9.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.9.2.0-preview\lib\net45\Microsoft.Azure.Management.Network.dll</HintPath>
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute.Test/Commands.Compute.Test.csproj]: @jobatzil 10.0.0-preview

File: src/ResourceManager/Compute/Commands.Compute.Test/packages.config [hunk 2]
```diff
@@ -7,7 +7,7 @@
   <package id="Microsoft.Azure.Graph.RBAC" version="3.2.0-preview" targetFramework="net45" />
   <package id="Microsoft.Azure.Management.Authorization" version="1.0.0" targetFramework="net45" />
   <package id="Microsoft.Azure.Management.Compute" version="14.0.0-prerelease" targetFramework="net45" />
-  <package id="Microsoft.Azure.Management.Network" version="9.1.0-preview" targetFramework="net45" />
+  <package id="Microsoft.Azure.Management.Network" version="9.2.0-preview" targetFramework="net45" />
```
Comment (cormacpayne) [inline src/ResourceManager/Compute/Commands.Compute.Test/packages.config]: @jobatzil 10.0.0-preview

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/AzureApplicationGatewayWebApplicationFirewallConfigurationBase.cs [hunk 3]
```diff
@@ -31,6 +33,24 @@ public class AzureApplicationGatewayWebApplicationFirewallConfigurationBase : Ne
...
+               Mandatory = true,
+               HelpMessage = "The type of the web application firewall rule set. Possible values are: 'OWASP'.")]
+        [ValidateNotNullOrEmpty]
+        public string RuleSetType { get; set; }
+
+        [Parameter(
+               Mandatory = true,
+               HelpMessage = "The version of the rule set type.")]
+        [ValidateNotNullOrEmpty]
+        public string RuleSetVersion { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/AzureApplicationGatewayWebApplicationFirewallConfigurationBase.cs]: @jobatzil adding parameters `RuleSetType` and `RuleSetVersion` that are mandatory is a breaking change - existing scripts that use these cmdlets will not work

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/NewAzureApplicationGatewayWebApplicationFirewallConfigurationCommand.cs [hunk 4]
```diff
@@ -17,24 +17,24 @@
 
 namespace Microsoft.Azure.Commands.Network
 {
-    [Cmdlet(VerbsCommon.New, "AzureRmApplicationGatewayWebApplicationFirewallConfiguration", SupportsShouldProcess = true),
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/NewAzureApplicationGatewayWebApplicationFirewallConfigurationCommand.cs]: @jobatzil why was this removed?

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/SetAzureApplicationGatewayWebApplicationFirewallConfigurationCommand.cs [hunk 5]
```diff
@@ -17,7 +17,7 @@
 
 namespace Microsoft.Azure.Commands.Network
 {
-    [Cmdlet(VerbsCommon.Set, "AzureRmApplicationGatewayWebApplicationFirewallConfiguration", SupportsShouldProcess = true),
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/SetAzureApplicationGatewayWebApplicationFirewallConfigurationCommand.cs]: @jobatzil why was this removed?

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallDisabledRuleGroup/AzureApplicationGatewayFirewallDisabledRuleGroupConfigBase.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        [Parameter(
+            Mandatory = true,
+            HelpMessage = "The name of the rule group that will be disabled.")]
+        [ValidateNotNullOrEmpty]
+        public string RuleGroupName { get; set; }
+
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "The list of rules that will be disabled. If null, all rules of the rule group will be disabled.")]
+        public List<int> Rules { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallDisabledRuleGroup/AzureApplicationGatewayFirewallDisabledRuleGroupConfigBase.cs]: @jobatzil according to [PowerShell *Strongly Encouraged Development Guidelines*](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx), parameter nouns should be singular:

_Avoid using plural names for parameters whose value is a single element. This includes parameters that take arrays or lists because the user might supply an array or list with only one element._

Let's think of a different noun that can be used instead of `Rules`. How does `RuleList` work?

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallDisabledRuleGroup/NewAzureApplicationGatewayFirewallDisabledRuleGroupConfigCommand.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Network.Models;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.New, "AzureRmApplicationGatewayFirewallDisabledRuleGroupConfig"),
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallDisabledRuleGroup/NewAzureApplicationGatewayFirewallDisabledRuleGroupConfigCommand.cs]: @jobatzil I'm worried that the name of this cmdlet is too long. Is there any way we can keep the name, but add an alias to something shorter?

File: src/ResourceManager/Network/Commands.Network/Models/PSApplicationGatewayAvailableWafRuleSetsResult.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Commands.Network.Models
+{
+    public class PSApplicationGatewayAvailableWafRuleSetsResult
+    {
+        public List<PSApplicationGatewayFirewallRuleSet> Value { get; set; }
+
+        [JsonIgnore]
+        public string ValueText
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/Models/PSApplicationGatewayAvailableWafRuleSetsResult.cs]: @jobatzil for each of the classes you are adding to the `Model` folder, can you make sure that each of their properties have corresponding help with them?

```cs
/// <summary>
/// Brief summary of property.
/// </summary>
public string ValueText
{
    ...
}

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/ApplicationGatewayTests.cs [hunk 9]
```diff
@@ -30,6 +30,13 @@ public ApplicationGatewayTests(ITestOutputHelper output)
 
         [Fact]
         [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void TestAvailableWafRuleSets()
+        {
+            NetworkResourcesController.NewInstance.RunPsTest(string.Format("Test-AvailableWafRuleSets -baseDir '{0}'", AppDomain.CurrentDomain.BaseDirectory));
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/ApplicationGatewayTests.cs]: @jobatzil was there a test recording file associated with this test? I didn't see one. If so, make sure it is added in the test csproj file.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/AzureApplicationGatewayWebApplicationFirewallConfigurationBase.cs (hunk 3 above)
Comment (jobatzil) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/AzureApplicationGatewayWebApplicationFirewallConfigurationBase.cs]: Yes, it is also mandatory in the new azure-sdk

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/NewAzureApplicationGatewayWebApplicationFirewallConfigurationCommand.cs (hunk 4 above)
Comment (jobatzil) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/NewAzureApplicationGatewayWebApplicationFirewallConfigurationCommand.cs]: it only changes the local object.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/SetAzureApplicationGatewayWebApplicationFirewallConfigurationCommand.cs (hunk 5 above)
Comment (jobatzil) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/SetAzureApplicationGatewayWebApplicationFirewallConfigurationCommand.cs]: same as above.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallDisabledRuleGroup/AzureApplicationGatewayFirewallDisabledRuleGroupConfigBase.cs (hunk 6 above)
Comment (jobatzil) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallDisabledRuleGroup/AzureApplicationGatewayFirewallDisabledRuleGroupConfigBase.cs]: Hmm, this is the name we use on every level (powershell, azure-ask, backend)

File: src/ResourceManager/Network/Commands.Network/Models/PSApplicationGatewayAvailableWafRuleSetsResult.cs (hunk 8 above)
Comment (jobatzil) [inline src/ResourceManager/Network/Commands.Network/Models/PSApplicationGatewayAvailableWafRuleSetsResult.cs]: We've never done that before, but I can do that.

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/ApplicationGatewayTests.cs (hunk 9 above)
Comment (jobatzil) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/ApplicationGatewayTests.cs]: yes, but it is not uploaded yet, I may change the cmdlet before it gets finalized.

Comment (jobatzil) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/ApplicationGatewayTests.cs]: I uploaded the test recordings.

**Comment by jobatzil** [comment]
@cormacpayne I didn't update the TestApplicationGatewayCRUD test recordings yet.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallDisabledRuleGroup/AzureApplicationGatewayFirewallDisabledRuleGroupConfigBase.cs (hunk 6 above)
Comment (jobatzil) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallDisabledRuleGroup/AzureApplicationGatewayFirewallDisabledRuleGroupConfigBase.cs]: I will leave it like that to keep it consistent.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallDisabledRuleGroup/NewAzureApplicationGatewayFirewallDisabledRuleGroupConfigCommand.cs (hunk 7 above)
Comment (jobatzil) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallDisabledRuleGroup/NewAzureApplicationGatewayFirewallDisabledRuleGroupConfigCommand.cs]: As discussed offline, there is no shorter name that still matches our current pattern. We are actually thinking of introducing shorter aliases for all of the cmdlets (agw instead of applicationGateway?), this will not be part of this PR though.
